README.md
# Page Template Example

An example WordPress plugin used to show how to include templates with your plugins and programmatically add them to the active theme.

You can read more about this plugin and how it works in [this blog post](http://tommcfarlin.com/wordpress-page-template-in-plugin/).

## How It Works

The plugin works like this:

1. Upon activation, it checks to see if a template with the same filename exists within the active theme's root directory.
2. If so, the plugin will not do anything; however, if no template file exists, then it will create the template in the root of the active theme.
3. Upon deactivation, the template will be removed from the active theme directory.

## Installation

### Using The WordPress Dashboard

1. Navigate to the 'Add New' Plugin Dashboard
2. Select `page-template-example.zip` from your computer
3. Upload
4. Activate the plugin on the WordPress Plugin Dashboard

### Using FTP

1. Extract `page-template-dashboard.zip` to your computer
2. Upload the `page-template-dashboard` directory to your `wp-content/plugins` directory
3. Activate the plugin on the WordPress Plugins Dashboard

## License

The WordPress Plugin Boilerplate is licensed under the GPL v2 or later.

> This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License, version 2, as
published by the Free Software Foundation.

> This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

> You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

## Known Issues

For information about changes related to the project, be sure to review the [ChangeLog](https://github.com/tommcfarlin/page-template-example/blob/master/ChangeLog.md).

## 1.0.0 (26 October 2013)

The templating mechanism has changed drastically in `1.0.0`. In order to add templates to your theme using this plugin,
you must do the following:

1. Review Lines 91 - 94 in `class-page-templates.php`.
2. Notice that we're defining templates in an associative array - the example templates exist in the `templates directory`.

To implement your own template, you must:

1. Add the template to the `templates directory`
2. Update the associated array to include the key/value pair necessary for updating the page attribute dropdown.

## 0.1.0 (7 March 2013)

Because this plugin is used strictly for example purposes, there are a number of things that are not properly handled:

* If the user changes a theme, the plugin does not move the template to the new theme's directory.
* There's practically no error handling for the file operations.
* If a template with the same file name already exists, there's no notification or error handling in this case so the user has no idea that the bundled template was not activsated.
* …I'm sure there are more.

This particular plugin is tagged as **0.1.0** and is open to any and all pull requests to help make this better, so feel free to contribute to resolve any of the above issues or any that you experience.

## Author Information

Page Template Example originally started and is maintained by [Tom McFarlin](http://twitter.com/tommcfarlin/), but is constantly under development thanks to the contributions from other WordPress developers.

ChangeLog.md
## ChangeLog

## 1.0.0 (26 October 2013)

* Adding a `ChangeLog`
* Updating the `README`
* Adding the `LICENSE` to the plugin
* Updating code comments for clarity
* Moving the plugin header to the core plugin file from the plugin's class
* Adding a second template to the plugin for additional demonstration purposes
* Adding a function for retrieving the plugin locale
* Updating the since tags in the documentation
* Adding a plugin slug attribute
* Adding a getter to retrieve the slug from the plugin
* Updating localization file to match the slug
* Adding a `changelog`
* Updating the documentation for the attributes
* Adding a constant to track the version of the plugin
* Updating the theme description and version
* Adding a core plugin file to instantiate the plugin on the `plugins_loaded` hook
* Renaming the `lang` directory to `languages`
* Renaming the core plugin file to match the name of the plugin
* Removed the file copy/delete methods and made it cache based. This removes the need to copy and delete files on activation. (Desertsnowman)

## 0.1.0 (7 March 2013)

* Initial release

page-template-example.php
<?php
/**
 * Page Template Example
 *
 * An example WordPress plugin used to show how to include templates with your
 * plugins and programmatically add them to the active theme
 *
 * @package   PTE
 * @author    Tom McFarlin <tom@tommcfarlin.com?
 * @license   GPL-2.0+
 * @link      https://github.com/tommcfarlin/page-template-example
 * @copyright 2013 Tom McFarlin
 *
 * @wordpress-plugin
 * Plugin Name:       Page Template Example
 * Plugin URI:        https://github.com/tommcfarlin/page-template-example
 * Description:       An example WordPress plugin used to show how to include templates with your plugins and programmatically add them to the active theme
 * Version:           1.0.0
 * Author:            Tom McFarlin
 * Author URI:        http://tommcfarlin.com
 * Text Domain:       pte-locale
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Domain Path:       /languages
 * GitHub Plugin URI: https://github.com/tommcfarlin/page-template-example/
 */

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
} // end if

require_once( plugin_dir_path( __FILE__ ) . 'class-page-template-example.php' );
add_action( 'plugins_loaded', array( 'Page_Template_Plugin', 'get_instance' ) );

class-page-template-example.php
<?php

/**
 * This plugin allows you to include templates with your plugin so that they can
 * be added with any theme.
 *
 * @package Page Template Example
 * @version 1.0.0
 * @since 	0.1.0
 */
class Page_Template_Plugin {

    /**
     * Plugin version, used for cache-busting of style and script file references.
     *
     * @since   1.0.0
     *
     * @var     string
     */
    const VERSION = '1.0.0';

    /**
     * Unique identifier for the plugin.
     *
     * The variable name is used as the text domain when internationalizing strings
     * of text.
     *
     * @since    1.0.0
     *
     * @var      string
     */
    protected $plugin_slug;

	/**
	 * A reference to an instance of this class.
	 *
	 * @since 0.1.0
	 *
	 * @var   Page_Template_Plugin
	 */
	private static $instance;

	/**
	 * The array of templates that this plugin tracks.
	 *
	 * @var      array
	 */
	protected $templates;


	/**
	 * Returns an instance of this class. An implementation of the singleton design pattern.
	 *
	 * @return   Page_Templae_Example    A reference to an instance of this class.
	 * @since    1.0.0
	 */
	public static function get_instance() {

		if( null == self::$instance ) {
			self::$instance = new Page_Template_Plugin();
		} // end if

		return self::$instance;

	} // end getInstance

	/**
	 * Initializes the plugin by setting localization, filters, and administration functions.
	 *
	 * @version		1.0.0
     * @since 		1.0.0
	 */
	private function __construct() {

		$this->templates = array();
		$this->plugin_locale = 'pte';

		// Grab the translations for the plugin
		add_action( 'init', array( $this, 'load_plugin_textdomain' ) );

		// Add a filter to the page attributes metabox to inject our template into the page template cache.
		add_filter('page_attributes_dropdown_pages_args', array( $this, 'register_project_templates' ) );

		// Add a filter to the save post in order to inject out template into the page cache
		add_filter('wp_insert_post_data', array( $this, 'register_project_templates' ) );

		// Add a filter to the template include in order to determine if the page has our template assigned and return it's path
		add_filter('template_include', array( $this, 'view_project_template') );

		// Add your templates to this array.
		$this->templates = array(
			'template-example.php'     => __( 'Example Page Template', $this->plugin_slug ),
			'template-example-two.php' => __( 'Example Page Template II', $this->plugin_slug )
		);


	} // end constructor

    /**
     * Load the plugin text domain for translation.
     *
     * @since    1.0.0
     */
    public function load_plugin_textdomain() {

	    $domain = $this->plugin_slug;
	    $locale = apply_filters( 'plugin_locale', get_locale(), $domain );

	    load_textdomain( $domain, trailingslashit( WP_LANG_DIR ) . $domain . '/' . $domain . '-' . $locale . '.mo' );
	    load_plugin_textdomain( $domain, FALSE, basename( dirname( __FILE__ ) ) . '/languages/' );

    } // end load_plugin_textdomain

	/**
	 * Adds our template to the pages cache in order to trick WordPress
	 * into thinking the template file exists where it doens't really exist.
	 *
	 * @param   array    $atts    The attributes for the page attributes dropdown
	 * @return  array    $atts    The attributes for the page attributes dropdown
	 * @verison	1.0.0
	 * @since	1.0.0
	 */
	public function register_project_templates( $atts ) {

		// Create the key used for the themes cache
		$cache_key = 'page_templates-' . md5( get_theme_root() . '/' . get_stylesheet() );

		// Retrieve the cache list. If it doesn't exist, or it's empty prepare an array
		$templates = wp_cache_get( $cache_key, 'themes' );
		if ( empty( $templates ) ) {
			$templates = array();
		} // end if

		// Since we've updated the cache, we need to delete the old cache
		wp_cache_delete( $cache_key , 'themes');

		// Now add our template to the list of templates by merging our templates
		// with the existing templates array from the cache.
		$templates = array_merge( $templates, $this->templates );

		// Add the modified cache to allow WordPress to pick it up for listing
		// available templates
		wp_cache_add( $cache_key, $templates, 'themes', 1800 );

		return $atts;

	} // end register_project_templates

	/**
	 * Checks if the template is assigned to the page
	 *
	 * @version	1.0.0
	 * @since	1.0.0
	 */
	public function view_project_template( $template ) {

		global $post;

		if ( ! isset( $this->templates[ get_post_meta( $post->ID, '_wp_page_template', true ) ] ) ) {
			return $template;
		} // end if

		$file = plugin_dir_path( __FILE__ ) . 'templates/' . get_post_meta( $post->ID, '_wp_page_template', true );

		// Just to be safe, we check if the file exist first
		if( file_exists( $file ) ) {
			return $file;
		} // end if

		return $template;

	} // end view_project_template


	/**
	 * Retrieves and returns the slug of this plugin. This function should be called on an instance
	 * of the plugin outside of this class.
	 *
	 * @return  string    The plugin's slug used in the locale.
	 * @version	1.0.0
	 * @since	1.0.0
	 */
	public function get_locale() {
		return $this->plugin_slug;
	} // end get_locale

} // end class


plugin.php
<?php
/*
Plugin Name: Page Template Plugin
Plugin URI: http://github.com/tommcfarlin/page-template-plugin/
Description: An example plugin used to show how to include page templates with your plugin.
Version: 0.1
Author: Tom McFarlin
Author URI: http://tommcfarlin.com/
Author Email: tom@tommcfarlin.com
License:

  Copyright 2013 Tom McFarlin (tom@tommcfarlin.com)

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License, version 2, as 
  published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
  
*/

if( ! defined( 'PAGE_TEMPLATE_PLUGIN' ) ) {
	define( 'PAGE_TEMPLATE_PLUGIN', '0.1' );
} // end if

/**
 * @package Page Template Example
 * @version 0.1
 * @since 	0.1
 */
class Page_Template_Plugin {

	/*--------------------------------------------*
	 * Attributes
	 *--------------------------------------------*/

	/** A reference to an instance of this class. **/
	private static $instance;

	/*--------------------------------------------*
	 * Constructor
	 *--------------------------------------------*/
	
	/**
	 * Returns an instance of this class. This is the Singleton design pattern.
	 * 
	 * @return OBJECT 	A reference to an instance of this class.
	 */
	public static function getInstance() {

		if( null == self::$instance ) {
			self::$instance = new Page_Template_Plugin();
		} // end if

		return self::$instance;

	} // end getInstance

	/**
	 * Initializes the plugin by setting localization, filters, and administration functions.
	 *
	 * @version		1.0
     * @since 		1.0
	 */
	private function __construct() {

		add_action( 'init', array( $this, 'plugin_textdomain' ) );

		register_activation_hook( __FILE__, array( $this, 'register_project_template' ) );
		register_deactivation_hook( __FILE__, array( $this, 'deregister_project_template' ) );
		
	} // end constructor

	/*--------------------------------------------*
	 * Localization
	 *--------------------------------------------*/

	/**
	 * Loads the plugin text domain for translation
	 *
	 * @version		1.0
     * @since 		1.0
	 */
	public function plugin_textdomain() {
		load_plugin_textdomain( 'pte', false, dirname( plugin_basename( __FILE__ ) ) . '/lang' );
	} // end plugin_textdomain

	/*--------------------------------------------*
	 * Activation and Deactivation Hooks
	 *--------------------------------------------*/

	/**
	 * Copies the template from the `views/templates` directory to the root of the active theme
	 * directory so that it can be applied to pages.
	 *
	 * @verison	1.0
	 * @since	1.0
	 */
	public function register_project_template() {

		// Get the template source and destination for copying from the plugin to the theme directory
		$template_destination = $this->get_template_destination();
		$template_source = $this->get_template_source();

		// Now actually copy the template file from the plugin to the destination
		$this->copy_page_template( $template_source, $template_destination );

	} // end register_project_template
	
	/** 
	 * Removes the template from the theme directory that was added during theme activation.
	 *
	 * @version	1.0
	 * since	1.0
	 */
	public function deregister_project_template() {
		
		// Get the path to the theme
		$theme_dir = get_stylesheet_directory();
		$template_path = $theme_dir . '/template-example.php';
		
		// If the template file is in the theme path, delete it.
		if( file_exists( $template_path ) ) {
			unlink( $template_path );
		} // end if
		
	} // end deregister_project_template

	/*--------------------------------------------*
	 * Helper Methods
	 *--------------------------------------------*/

	/**
	 * @return string The destination to the plugin directory relative to the currently active theme
	 */
	private function get_template_destination() {
		return get_stylesheet_directory() . '/template-example.php';
	} // end get_template_destination

	/**
	 * @return string The path to the template file relative to the plugin.
	 */
	private function get_template_source() {
		return dirname( __FILE__ ) . '/templates/template-example.php';
	} // end get_template_source

	/**
	 * Copies the contents of the template from the source file in the plugin
	 * to the destination in the current theme directory.
	 * 
	 * This works by first creating an empty file, reading the contents of the template file
	 * then writing it into the empty file in the theme directory.
	 *
	 * Note that this version is for demonstration purposes only and does not include proper
	 * exception handling for when file operations fail.
	 * 
	 * @param  string $template_source      The path on which the template resides.
	 * @param  string $template_destination The path to which the template should be written.
	 */
	private function copy_page_template( $template_source, $template_destination ) {

		// After that, check to see if the template already exists. If so don't copy it; otherwise, copy if
		if( ! file_exists( $template_destination ) ) {
			
			// Create an empty version of the file
			touch( $template_destination );
			
			// Read the source file starting from the beginning of the file
			if( null != ( $template_handle = @fopen( $template_source, 'r' ) ) ) {
			
				// Now read the contents of the file into a string. Read up to the length of the source file
				if( null != ( $template_content = fread( $template_handle, filesize( $template_source ) ) ) ) {
				
					// Relinquish the resource
					fclose( $template_handle );
					
				} // end if
				
			} // end if
						
			// Now open the file for reading and writing
			if( null != ( $template_handle = @fopen( $template_destination, 'r+' ) ) ) {
			
				// Attempt to write the contents of the string
				if( null != fwrite( $template_handle, $template_content, strlen( $template_content ) ) ) {
				
					// Relinquish the resource
					fclose( $template_handle );
					
				} // end if

			} // end if
			
		} // end if

	} // end copy_page_template

} // end class

$GLOBALS['ptb'] = Page_Template_Plugin::getInstance();
